### PR TITLE
Side effects in BaseQNode._check_circuit removed.

### PR DESCRIPTION
**Pull request by dlasecki**
**Context:**
BaseQNode._check_circuit contains side effects.

**Description of the Change:**
Extracted the logic that causes side effects to a separate (private) method. Note that no unit tests are introduced since one does not unit test private methods.

**Benefits:**
Solves [#620](https://github.com/PennyLaneAI/pennylane/issues/630)

**Possible Drawbacks:**
This is just a refactor, hence no functional differences are introduced. The codebase should be searched for places where `BaseQNode._check_circuit` and the newly introduced `BaseQNode._get_decomposed_and_extended_queue` are used together. Perhaps, using `BaseQNode._get_decomposed_and_extended_queue` after `BaseQNode._check_circuit` is not necessary in certain places. Making this decision requires some more familiarity with the codebase.

**Related GitHub Issues:**
[#620](https://github.com/PennyLaneAI/pennylane/issues/630)

### CODE REVIEW

File: pennylane/qnodes/base.py [hunk 1]
```diff
@@ -574,6 +574,8 @@ def _construct(self, args, kwargs):
 
         # check the validity of the circuit
         self._check_circuit(res)
+        # set self.ops to a decomposed queue appended with an output res returned by the quantum function
+        self.ops = self._get_decomposed_and_extended_queue(res)
```
Comment (josh146) [inline pennylane/qnodes/base.py:578]: It's now much clearer how `ops` comes into being!

**Review by josh146 — COMMENTED** [review]
It's a small change @dlasecki, but it leads to a significant improvement in code readability and maintainability!

Even though `_get_decomposed_and_extended_queue` is a private function, it would be great to see an example of a unit test to go along with it.

**Comment by dlasecki** [comment]
> It's a small change @dlasecki, but it leads to a significant improvement in code readability and maintainability!
> 
> Even though `_get_decomposed_and_extended_queue` is a private function, it would be great to see an example of a unit test to go along with it.

@josh146 Thanks for feedback! Sure, just give me few more days since I am pretty sick right now.

**Comment by josh146** [comment]
No worries, @dlasecki, hope you feel better :)

**Comment by co9olguy** [comment]
Thanks @dlasecki for adding the new tests!

**Comment by josh146** [comment]
Closing this PR for now, as `BaseQNode` is deprecated and will be replaced by the new `QNode` class

**Pull request closed without merging**